### DISCUSSION TITLE
Reducing support for Internet Explorer (IE) in GOV.UK Frontend

### DISCUSSION THREAD

**Discussion by querkmachine** [alphagov/govuk-frontend#2607] [Announcements]
**Closing date:** Closed - deadline was 13th May 2022

**Status:** Accepted. You can follow implementation progress [here](https://github.com/alphagov/govuk-frontend/issues/2621).

## Scope

This proposal is:

- only about browser support in GOV.UK Frontend
- dependent on an expected change to [the Service Manual to reduce IE11 support](https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices) from ‘compliant’ to ‘functional’ or remove it entirely

## Summary

We're anticipating the requirement to test in Internet Explorer 11 (IE11) will be updated in the Service Manual within the next year. When that occurs, we plan to reduce our level of support in GOV.UK Frontend for all IE versions.

Currently, GOV.UK Frontend provides:

- a 'compliant' experience for IE11, which means our components and styles look and behave the same in IE11 and more modern browsers.

- 'functional' support for IE9 and IE10, which means our components are usable in those browsers, but may not look or behave the same.

- 'functional' support for IE8, through the use of [an optional, additional, stylesheet](https://frontend.design-system.service.gov.uk/supporting-ie8/).

We intend to reduce the level of support for IE11 from 'compliant' to 'functional' in a future major release of GOV.UK Frontend. In doing so, we will stop shipping JavaScript functionality for IE.

At the same time, we intend to end support for IE versions 8, 9 and 10.

We'd like your feedback on these proposed changes and how they might affect your services. You can:

💬 add comments to this discussion
✉️ send feedback to govuk-design-system-support@digital.cabinet-office.gov.uk
⬆️ Upvote this comment if you agree with the proposed changes and have no other feedback to give

## Background

From 15 June 2022, [Microsoft will end support for IE11 and begin removing it from Windows](https://techcommunity.microsoft.com/t5/windows-it-pro-blog/internet-explorer-11-desktop-app-retirement-faq/ba-p/2366549). Users will be redirected to Microsoft Edge instead.

Mainstream support for IE8 and IE9 ended in January 2016, and for IE10 in January 2020. This included an end to security patches for new vulnerabilities that became known after these dates.

We’re making this decision based on data from multiple sources. Analytics data, reported by GOV.UK and other service teams, show that visits using IE11 have dropped to, on average, below 1%. Users of older IE versions make up less than 0.01% of visits to GOV.UK as of February this year.

The [latest WebAIM screenreader survey](https://webaim.org/projects/screenreadersurvey9/) reports that usage of IE with screen readers has dropped from 70% in 2009 (IE11) to 3.3% (all IE versions) in 2021. JAWS and IE11, a previously popular screen reader-browser combination, has dropped from 24.7% to 1.9%.

We have a hunch that some users of assistive technologies may find it hard to switch from IE to a different browser. Some analytics data also suggest that specific teams see a higher percentage of visits from IE11. For example, due to restrictions enforced by organisation IT policies. If you have either of these issues, we would especially like to hear from you so we can learn more. Likewise, if you have concerns about how this change may affect particular user groups.

## Problem

We've previously estimated that about 50% of the JavaScript we ship is made up of polyfills that most of our users do not need. We include these polyfills to provide support for IE, but they're sent to all browsers whether they're needed or not.

Polyfills do not scale well when we think about building more complex components. For example, modals, which may rely on modern browser features. And as we've seen from analytics data, we're shipping a large percentage of code for a shrinking proportion of users.

As well as polyfills, GOV.UK Frontend includes numerous CSS adjustments that are specific to older versions of IE, particularly IE8. Like polyfills, many of these adjustments are shipped to all browsers whether they're needed or not.

Making sure that new components, feature updates and bug fixes will work in IE requires more of our time and resources than with other browsers. Keeping the current level of support seems unwise, given the small number of IE users accessing government services.

The requirement to support IE at the 'compliant' level also prevents us from making use of more efficient CSS and JavaScript features, and makes it difficult to introduce more complex components.

The changes we're proposing will mean we can remove these adjustments, which will:

- improve performance for everyone
- make it easier for us to introduce more complex components in the future
- free up resources to fix issues and deliver features that benefit a wider user base

## Proposal

We'll publish a major version of GOV.UK Frontend (v5.0.0) that reduces the level of support for IE11 to 'functional', and removes support for IE10 and below. This version will include the removal of code that is specific to older versions of IE, such as polyfills and the IE8-specific stylesheet. 

If you update to v5, your service will not stop working on IE11. However, components and styles may look and behave differently, as they will no longer use JavaScript.

Service teams should be [using a progressive enhancement approach](https://www.gov.uk/service-manual/technology/using-progressive-enhancement) to make sure users can still access any content and complete their tasks. In the future, we may make further changes that affect the look and behaviour of components and styles in IE. However, we will not treat them as breaking changes.

If you need to continue providing a ‘compliant’ experience in IE11, or still need to support older versions of IE, you should keep using [GOV.UK Frontend version 4](https://github.com/alphagov/govuk-frontend/releases/tag/v4.0.0).

We’ll continue to provide minimal support for v4 for the next 12 months by publishing urgent security updates and major bug fixes. We will not add any new components to v4, or update it to meet Web Content Accessibility Guidelines (WCAG) 2.2. After 12 months, we’ll revisit the data and decide whether we need to extend the period of support.

As part of this work, we’ll be considering how to continue providing documentation for users who still need to use v4 to support IE. If this sounds like you, we would especially like to hear from you, and also if there’s anything else we should consider.

**Comment by robertdeniszczyc2** [comment]
Hello,

Does the "Functional" support definition need to account for IE11 Mode in Edge, which [will be supported until at least 2029](https://techcommunity.microsoft.com/t5/windows-it-pro-blog/internet-explorer-11-desktop-app-retirement-faq/ba-p/2366549)?

This might be more a question for the Service Manual team, but I think it's still relevant here, as is there a consideration or requirement  that IE11 Mode will need be supported by govuk-frontend in a Functional manner for at least another 7+ years, while it's running inside  a browser we are Compliant with?

From what I understand IE11 mode in Edge runs Trident MSHTML engine under the hood so it's more than an emulator, but I believe there are still some differences when comparing IE11 "proper" to IE11 mode in Edge.

Cheers,
Rob

**Reply by querkmachine** [reply]
Hi Rob, 

Thanks for the question! 

Our understanding is that [IE Mode is intended to render pages in exactly the same was as "proper" IE11](https://docs.microsoft.com/en-gb/deployedge/edge-ie-mode#ie-mode-supports-the-following-internet-explorer-functionality), which is why we haven't considered it separately for the purposes of future support.   

In the event that IE Mode does deviate, we will probably favour supporting IE11 over IE Mode, as IE Mode doesn't ship with any built-in debugging tools—making it much harder for us to fix any issues that should arise.

**Reply by robertdeniszczyc2** [reply]
Thanks @querkmachine that makes sense

Cheers,
Rob

**Comment by frankieroberto** [comment]
👍  from me. Does this mean you would then consider switching to using ECMAScript 2015 (ES6)?

**Reply by querkmachine** [reply]
Hey Frankie, 

Initially we will only be removing IE polyfills, and not significantly changing the existing JavaScript. Going forward, we will start to refactor and incorporate features introduced in ES2015 and newer specifications, and these wouldn't be considered breaking changes. 

We probably won't be using 'bleeding edge' JS features—as we want to avoid transpiling and creating a significant disconnect between the code we author and the code we ship, or having to re-introduce polyfills for other browsers—instead favouring JS features that have been well supported for many years, so that there is less detriment to the long tail of users stuck on older Safari or Chrome versions [for whatever reasons](https://blog.jim-nielsen.com/2022/a-web-for-all/).

Hope that answers your question!

**Reply by frankieroberto** [reply]
Sounds sensible! I'm not a huge fan of doing loads of transpiling either. 👍

**Comment by alexbishop1** [comment]
As I understand it, Microsoft is only ending support for IE11 on Windows 10. Other operating systems are unaffected. This includes [Windows 8.1, which is still supported until 10 January 2023](https://docs.microsoft.com/en-gb/lifecycle/products/windows-81). Although Edge is available for Windows 8.1, I cannot recall if it was ever pushed out to users who did not choose to install it, or if Windows 8.1 users prone to inertia will still only have IE.

**Reply by querkmachine** [reply]
Hi Alex,

Thanks for the info! 

We're making these decisions based on Microsoft's mainstream support end dates rather than their extended support dates, as extended support can continue for many, many years after the browser has otherwise fallen out of use. [Internet Explorer 7 is still supported by one versions of Windows Embedded](https://docs.microsoft.com/en-us/lifecycle/faq/internet-explorer-microsoft-edge#what-is-the-lifecycle-policy-for-internet-explorer-), for example.

It's a good point though! Our prior figures are only based on what IE versions are in use, not which operating system they're being used on. I'll try and find out what proportion of IE11 users are doing so from Windows 8.1.

**Reply by querkmachine** [reply]
Hi Alex,

These were the stats on IE11 usage on GOV.UK last month:

|Operating system|% of sessions|Sessions (thousands)|
|:-|-:|-:|
|Windows 10|82.2|564.9|
|Windows 8.1|5.2|35.6|
|Windows 8|4.6|31.5|
|Windows 7|8.0|55.3|

Although these are big numbers when compared to one another, IE11 users on Windows 8.1 and below only accounted for 0.001% of total sessions on GOV.UK.

**Comment by sdh100shaun** [comment]
👍  this would be very welcome.

**Comment by stevenjmesser** [comment]
Microsoft permanently disabled Internet Explorer 11 on versions of Windows 10 [this week](https://pxlnv.com/linklog/internet-explorer-killswitch/). We've seen a ~28% reduction in sessions using IE11 since 30 January 2023 but will keep monitoring this in case it's due to half term.
